### PR TITLE
Implement topk (k!=1) in token choice

### PR DESCRIPTION
**Pull request by maciek-pioro**
# Description
Title says it all.

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] [I attached link to neptune if it was necessary](https://app.neptune.ai/o/pmtest/org/llm-random/runs/details?viewId=9a9e4902-9b67-478b-adb4-36df9e86a9c1&detailsTab=metadata&shortId=LLMRANDOM-5853&type=run&query=(sys%2Fid%3Astring%20%3D%20%22LLMRANDOM-5853%22)&sortBy=%5B%22sys%2Fcreation_time%22%5D&sortFieldType=%5B%22datetime%22%5D&sortFieldAggregationMode=%5B%22auto%22%5D&sortDirection=%5B%22descending%22%5D&suggestionsEnabled=true&lbViewUnpacked=true)
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [ ] ~If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters~
   - [ ] ~If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)~
   - [ ] ~If there is a need for a second reviewer I requested additional review~

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: research/conditional/moe_layers/token_choice.py [hunk 1]
```diff
@@ -28,6 +29,7 @@ def __init__(
         self.load_balancing_loss_weight = load_balancing_loss_weight
         self.use_einsum = use_einsum
         self.dmodel = dmodel
+        self.experts_per_token = experts_per_token
```
Comment (Simontwice) [inline research/conditional/moe_layers/token_choice.py:32]: maybe sth like `routing_top_k` would be more standard

File: research/conditional/moe_layers/token_choice.py [hunk 2]
```diff
@@ -124,20 +139,38 @@ def choose_expert(self, gate_out) -> tuple[torch.Tensor, torch.Tensor]:
...
+            ),  # move
+            "gate_softmax_all_values": make_histogram(
+                self.logging_cache["gate_softmax_all_values"].flatten()  # move
+            ),
+            "tokens_per_expert_counts": make_histogram(
+                self.logging_cache["tokens_per_expert"]
+            ),
+            "load_balancing_loss": self.logging_cache["load_balancing_loss"],
+            "dropped_tokens_ratio": self.logging_cache["dropped_tokens_ratio"],
+        }
```
Comment (Simontwice) [inline research/conditional/moe_layers/token_choice.py:172]: does this work? I've had problems before. I'd love to see a neptune run with `logging_interval_heavy` set to like 10 or sth.

File: research/conditional/moe_layers/token_choice.py [hunk 3]
```diff
@@ -223,37 +258,19 @@ def forward(self, x: torch.Tensor):
...
                     i, : len(indices_of_tokens_for_expert[i])
-                ]
-
-        # multiply output by softmax values
-        with measure_time(self, "multiply_output_by_softmax"):
-            if self.use_einsum:
-                output = einsum(
-                    "n_tokens dmodel, n_tokens -> n_tokens dmodel", output, expert_gate
+                ] * masked_expert_gate[indices_of_tokens_for_expert[i], i].unsqueeze(
+                    dim=1
```
Comment (Simontwice) [inline research/conditional/moe_layers/token_choice.py:267]: thats some involved code. Is that decomposable into multiple steps, perhaps?

Comment (maciek-pioro) [reply research/conditional/moe_layers/token_choice.py]: I made it a bit more readabloe

**Review by Simontwice — APPROVED** [review]
LGTM, logging would be nice but is not crucial rn; please do leave a reminder in the code that it needs fixing if FSDP is used.

File: research/conditional/moe_layers/token_choice.py (hunk 2 above)
Comment (maciek-pioro) [reply research/conditional/moe_layers/token_choice.py]: Next PR

**Pull request merged**